File: app/csharp/AccountingSystem.Tests/Services/BalanceServiceTest.cs
using System.Data;
using AccountingSystem.Application.Services;
using AccountingSystem.Infrastructure.Persistence.Repositories;
using Dapper;
using FluentAssertions;
using Npgsql;
using Xunit;

namespace AccountingSystem.Tests.Services;

/// <summary>
/// 残高管理サービスのテスト
/// </summary>
public class BalanceServiceTest : DatabaseTestBase
{
    static BalanceServiceTest()
    {
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
    }

    [Fact(DisplayName = "日次残高をUPSERTで更新できる - 新規登録")]
    public async Task Test_日次残高をUPSERTで更新できる_新規登録()
    {
        // Given: 勘定科目を事前登録
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();

        var accountCode = "1020";
        await connection.ExecuteAsync(@"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分""
            ) VALUES (@Code, '普通預金', '資産', 'B', '1')
            ON CONFLICT (""勘定科目コード"") DO NOTHING",
            new { Code = accountCode });

        var service = new BalanceService(ConnectionString);
        var entryDate = new DateOnly(2025, 1, 15);

        // When: 日次残高を更新（新規登録）
        await service.UpdateDailyBalanceAsync(
            entryDate,
            accountCode,
            subAccountCode: "",
            departmentCode: "",
            projectCode: "",
            settlementFlag: 0,
            debitAmount: 100000.00m,
            creditAmount: 0.00m);

        // Then: 残高が登録されている
        var allRows = await connection.QueryAsync<dynamic>(@"SELECT * FROM ""日次勘定科目残高""");
        allRows.Should().HaveCount(1);

        var balance = allRows.First();
        ((decimal)balance.借方金額).Should().Be(100000.00m);
        ((decimal)balance.貸方金額).Should().Be(0.00m);
    }

    [Fact(DisplayName = "日次残高をUPSERTで更新できる - 既存レコードに加算")]
    public async Task Test_日次残高をUPSERTで更新できる_既存レコードに加算()
    {
        // Given: 勘定科目と既存の日次残高を事前登録
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();

        var accountCode = "1021";
        await connection.ExecuteAsync(@"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分""
            ) VALUES (@Code, 'テスト科目', '資産', 'B', '1')
            ON CONFLICT (""勘定科目コード"") DO NOTHING",
            new { Code = accountCode });

        var entryDate = new DateOnly(2025, 1, 15);
        await connection.ExecuteAsync(@"
            INSERT INTO ""日次勘定科目残高"" (
                ""起票日"", ""勘定科目コード"", ""補助科目コード"", ""部門コード"",
                ""プロジェクトコード"", ""決算仕訳フラグ"", ""借方金額"", ""貸方金額""
            ) VALUES (@EntryDate, @AccountCode, '', '', '', 0, 100000.00, 0.00)",
            new { EntryDate = entryDate, AccountCode = accountCode });

        var service = new BalanceService(ConnectionString);

        // When: 同じキーで日次残高を更新（既存に加算）
        await service.UpdateDailyBalanceAsync(
            entryDate,
            accountCode,
            subAccountCode: "",
            departmentCode: "",
            projectCode: "",
            settlementFlag: 0,
            debitAmount: 50000.00m,
            creditAmount: 30000.00m);

        // Then: 残高が加算されている
        var allRows = await connection.QueryAsync<dynamic>(@"SELECT * FROM ""日次勘定科目残高""");
        allRows.Should().HaveCount(1);

        var balance = allRows.First();
        ((decimal)balance.借方金額).Should().Be(150000.00m); // 100000 + 50000
        ((decimal)balance.貸方金額).Should().Be(30000.00m);   // 0 + 30000
    }

    [Fact(DisplayName = "部門別の日次残高を更新できる")]
    public async Task Test_部門別の日次残高を更新できる()
    {
        // Given: 勘定科目を事前登録
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();

        var accountCode = "4010";
        await connection.ExecuteAsync(@"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分""
            ) VALUES (@Code, '売上高', '収益', 'P', '4')
            ON CONFLICT (""勘定科目コード"") DO NOTHING",
            new { Code = accountCode });

        var service = new BalanceService(ConnectionString);
        var entryDate = new DateOnly(2025, 1, 15);

        // When: 部門別に日次残高を更新
        await service.UpdateDailyBalanceAsync(
            entryDate, accountCode, "", "001", "", 0, 0.00m, 300000.00m);
        await service.UpdateDailyBalanceAsync(
            entryDate, accountCode, "", "002", "", 0, 0.00m, 200000.00m);

        // Then: 部門ごとに別々の残高が登録されている
        var balances = await connection.QueryAsync<dynamic>(@"
            SELECT * FROM ""日次勘定科目残高""
            WHERE ""勘定科目コード"" = @AccountCode
            ORDER BY ""部門コード""",
            new { AccountCode = accountCode });

        var list = balances.ToList();
        list.Should().HaveCount(2);
        ((string)list[0].部門コード).Should().Be("001");
        ((decimal)list[0].貸方金額).Should().Be(300000.00m);
        ((string)list[1].部門コード).Should().Be("002");
        ((decimal)list[1].貸方金額).Should().Be(200000.00m);
    }
}

File: app/fsharp/AccountingSystem.Tests/Application/BalanceServiceTest.fs
namespace AccountingSystem.Tests.Application

open System
open System.Threading.Tasks
open Xunit
open FsUnit.Xunit
open AccountingSystem.Application.Repositories
open AccountingSystem.Application.Services

/// <summary>
/// IBalanceRepository のモック実装
/// </summary>
type MockBalanceRepository() =
    let mutable dailyBalanceUpdateCalls: DailyBalanceUpdateParams list = []
    let mutable journalItemUpdateCalls: string list = []
    let mutable monthlyConsolidateCalls: MonthlyBalanceConsolidateParams list = []

    /// 日次残高更新の呼び出し履歴を取得
    member this.DailyBalanceUpdateCalls = dailyBalanceUpdateCalls

    /// 仕訳明細からの更新呼び出し履歴を取得
    member this.JournalItemUpdateCalls = journalItemUpdateCalls

    /// 月次集計の呼び出し履歴を取得
    member this.MonthlyConsolidateCalls = monthlyConsolidateCalls

    interface IBalanceRepository with
        member this.UpdateDailyBalanceAsync(param: DailyBalanceUpdateParams) : Task<unit> =
            dailyBalanceUpdateCalls <- dailyBalanceUpdateCalls @ [param]
            Task.FromResult(())

        member this.UpdateBalanceFromJournalItemsAsync(journalNo: string) : Task<unit> =
            journalItemUpdateCalls <- journalItemUpdateCalls @ [journalNo]
            Task.FromResult(())

        member this.ConsolidateMonthlyBalanceAsync(param: MonthlyBalanceConsolidateParams) : Task<unit> =
            monthlyConsolidateCalls <- monthlyConsolidateCalls @ [param]
            Task.FromResult(())

/// <summary>
/// BalanceService のテスト
/// </summary>
type BalanceServiceTest() =

    [<Fact>]
    member this.``UpdateDailyBalanceAsync でリポジトリが正しいパラメータで呼び出される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)
            let entryDate = DateTime(2025, 1, 15)
            let accountCode = "1020"
            let subAccountCode = Some "001"
            let departmentCode = Some "SALES"
            let projectCode = Some "P001"
            let settlementFlag = Some 0
            let debitAmount = 100000.00M
            let creditAmount = 0.00M

            // When
            do! service.UpdateDailyBalanceAsync(
                entryDate,
                accountCode,
                subAccountCode,
                departmentCode,
                projectCode,
                settlementFlag,
                debitAmount,
                creditAmount)

            // Then
            mockRepository.DailyBalanceUpdateCalls.Length |> should equal 1
            let call = mockRepository.DailyBalanceUpdateCalls.[0]
            call.EntryDate |> should equal entryDate
            call.AccountCode |> should equal accountCode
            call.SubAccountCode |> should equal subAccountCode
            call.DepartmentCode |> should equal departmentCode
            call.ProjectCode |> should equal projectCode
            call.SettlementFlag |> should equal settlementFlag
            call.DebitAmount |> should equal debitAmount
            call.CreditAmount |> should equal creditAmount
        }

    [<Fact>]
    member this.``UpdateDailyBalanceAsync でオプション値が None の場合も正しく処理される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)
            let entryDate = DateTime(2025, 1, 15)
            let accountCode = "1020"

            // When
            do! service.UpdateDailyBalanceAsync(
                entryDate,
                accountCode,
                None,
                None,
                None,
                None,
                50000.00M,
                0.00M)

            // Then
            mockRepository.DailyBalanceUpdateCalls.Length |> should equal 1
            let call = mockRepository.DailyBalanceUpdateCalls.[0]
            call.SubAccountCode |> should equal None
            call.DepartmentCode |> should equal None
            call.ProjectCode |> should equal None
            call.SettlementFlag |> should equal None
        }

    [<Fact>]
    member this.``UpdateBalanceFromJournalItemsAsync でリポジトリが正しい仕訳番号で呼び出される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)
            let journalNo = "JV-2025-00001"

            // When
            do! service.UpdateBalanceFromJournalItemsAsync(journalNo)

            // Then
            mockRepository.JournalItemUpdateCalls.Length |> should equal 1
            mockRepository.JournalItemUpdateCalls.[0] |> should equal journalNo
        }

    [<Fact>]
    member this.``ConsolidateMonthlyBalanceAsync でリポジトリが正しいパラメータで呼び出される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)
            let fiscalYear = 2025
            let month = 1

            // When
            do! service.ConsolidateMonthlyBalanceAsync(fiscalYear, month)

            // Then
            mockRepository.MonthlyConsolidateCalls.Length |> should equal 1
            let call = mockRepository.MonthlyConsolidateCalls.[0]
            call.FiscalYear |> should equal fiscalYear
            call.Month |> should equal month
        }

    [<Fact>]
    member this.``複数回の呼び出しが正しく記録される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)

            // When: 複数回呼び出し
            do! service.UpdateBalanceFromJournalItemsAsync("JV-001")
            do! service.UpdateBalanceFromJournalItemsAsync("JV-002")
            do! service.UpdateBalanceFromJournalItemsAsync("JV-003")

            // Then
            mockRepository.JournalItemUpdateCalls.Length |> should equal 3
            mockRepository.JournalItemUpdateCalls.[0] |> should equal "JV-001"
            mockRepository.JournalItemUpdateCalls.[1] |> should equal "JV-002"
            mockRepository.JournalItemUpdateCalls.[2] |> should equal "JV-003"
        }

    [<Fact>]
    member this.``異なるメソッドの呼び出しが独立して記録される``() =
        task {
            // Given
            let mockRepository = MockBalanceRepository()
            let service = BalanceService(mockRepository)

            // When: 異なるメソッドを呼び出し
            do! service.UpdateDailyBalanceAsync(
                DateTime(2025, 1, 15),
                "1020",
                None,
                None,
                None,
                None,
                100000.00M,
                0.00M)
            do! service.UpdateBalanceFromJournalItemsAsync("JV-001")
            do! service.ConsolidateMonthlyBalanceAsync(2025, 1)

            // Then: 各メソッドの呼び出しが独立して記録されている
            mockRepository.DailyBalanceUpdateCalls.Length |> should equal 1
            mockRepository.JournalItemUpdateCalls.Length |> should equal 1
            mockRepository.MonthlyConsolidateCalls.Length |> should equal 1
        }
